Add separate cuda and nvrtc initialization, refactor generator code

brechtvl opened the pull request:
These new functions can be called to use nvrtc without CUDA hardware available, since `cuewInit()` can fail if that's not the case.

Changing the generator code to add these was a pain though. Instead I've made that code use cuew.template.c/h files. These are easier to edit, inserting autogenerated code in a few places rather than generating all C code through Python.

brechtvl committed:
Use cuew.template.c/h files to generate code

These are easier to edit for changes like the next commit, inserting
autogenerated code in a few places rather than generating all C code
through Python.

brechtvl commented:
I forgot to mention: this also updates the nvrtc dll names, and part of this patch was written by @LazyDodo.

sergeyvfx commented:
Part related on the template is something which will become handy in the future.

But part of split `cuewInit()` and `cuewNvrtcInit()` i didn't quite understand. We don't require hardware, and only requirement i see is CUDA driver version to be 4.0 at least. I don't think older drivers had NVRCT in them, and how is it possible that you have NVRTC without newer driver installed? Has it became part of display driver instead?

To me it seems we should have single `cuewInit()` function which will initialize all available wranglers. If it is REALLY needed to specify which parts we want to ignore, make it a bitmask argument to it.

Put all this information into a commit message. Having such a brief oneliners do not help at all. Also, don't use fullstop in commit subject (first line of commit message).

sergeyvfx reviewed (changes requested):
More information needed, see the comment in conversation.

brechtvl commented:
`libcuda` is part of the driver, and `dlopen()` on it fails there is no CUDA hardware. And `libnvrtc` can be installed as part of the CUDA toolkit, without the driver. What we are interested in is loading `libnvrtc` to compile programs without having any CUDA hardware.

If we add a argument to `cuewInit()`, we break backwards compatibility. I wouldn't mind doing that though, because I think it's bad that we are now automatically loading `libnvrtc` along with `libcuda`. Most of the time there is no reason to have the time and memory overhead of loading the compiler when all you need to do is render something.

sergeyvfx commented:
> What we are interested is loading libnvrtc to compile programs without having any CUDA hardware.

What is the usecase for that?

> If we add a argument to cuewInit(), we break backwards compatibility.

You break it anyway by removing NVRTC library load from `cuewInit()`. So everyone will need to change their programs anyway. I also not sure we ever declared API stable. We didn't even make a release or anything..

If you are really worried about compatibility, it's possible to keep compatibility by doing some variadic macro, which will convert `cuewInit()` to a call with both CUDA and NVRTC requested for load. But i don't think it worth it.

> Most of the time there is no reason to have the time and memory overhead of loading the compiler when all you need to do is render something.

Did you quantify memory and time? Is it really that bad?

brechtvl committed:
Update to new nvrtc library names on Windows

brechtvl committed:
Add cuewNvrtcVersion()

brechtvl committed:
Add cuewInit() flags to initialize cuda and nvrtc libraries separately

libcuda is part of the driver, and loading it can fail if there is no
CUDA hardware available. libnvrtc can be installed as part of the CUDA
toolkit without a driver, and can be used to compile programs on build
machines that do not have CUDA hardware.

On the other hand, it is also useful to avoid the overhead of loading
libnvrtc when only libcuda is needed.

brechtvl committed:
Update major version number for backwards compatibility breakage

brechtvl commented:
> What is the usecase for that?

Compiling CUDA binaries for Blender releases: https://developer.blender.org/D3045

> Did you quantify memory and time? Is it really that bad?

No, I did not quantify it, and it may not be bad in practice. But it's a big library, it might be doing a bunch of static initialization, and these things can change over time. It's just an argument in favor or breaking backwards compatibility, which you already seem to agree with so it doesn't matter.

sergeyvfx commented:
> Compiling CUDA binaries for Blender releases: https://developer.blender.org/D3045

Makes sense. But all such usecases are to be reflected in clear explanatory commit message.

> No, I did not quantify it, and it may not be bad in practice. But it's a big library, it might be doing a bunch of static initialization, and these things can change over time. It's just an argument in favor or breaking backwards compatibility, which you already seem to agree with so it doesn't matter.

Yes, make it a bitmask argument to cuewInit(). Internally you can still split the function up, to avoid too long of a function which does everything.

brechtvl commented:
Updated with flags for `cuewInit()`, version bump to indicate compatibility breakage, and better commit messages.

sergeyvfx commented on auto/cuew.template.c:
```diff
@@ -156,15 +145,87 @@ int cuewInit(void) {
+  }
+}
+
+int cuewNvrtcInit(void) {
```
Seems to be private function which is not exposed to a public API. Hence should be using static qualifier.

brechtvl commented:
Made function static.

sergeyvfx reviewed (approved):
Think we are good now.

Do you have commit access here?

brechtvl commented:
No, I wouldn't mind getting it though.
